### ISSUE TITLE
Pot sync not finding the last salt cracked in a forked session.

### ISSUE THREAD

**Issue by claudioandre** [magnumripper/JohnTheRipper#1821]
While building the salt hash table in ldr_sort_salts() - loader.c, the for loop code misses the last salt in the list.

I mean ```db->salt_hash[hash] = s``` is not happening for the last item.
****
A patch that fixes the problem and shows what is wrong.

```
diff --git a/src/loader.c b/src/loader.c
index bf9d20e..3cea7b0 100644
--- a/src/loader.c
+++ b/src/loader.c
@@ -1236,7 +1236,7 @@ static void ldr_sort_salts(struct db_main *db)
        /* finally, we re-build the linked list of salts */
        db->salts = ar[0].p;
        s = db->salts;
-       for (i = 1; i < db->salt_count; ++i) {
+       for (i = 1; i <= db->salt_count; ++i) {
                /* Rebuild salt hash table, if we still had one */
                if (db->salt_hash) {
                        int hash;
@@ -1245,8 +1245,10 @@ static void ldr_sort_salts(struct db_main *db)
                        if (!db->salt_hash[hash])
                                db->salt_hash[hash] = s;
                }
-               s->next = ar[i].p;
-               s = s->next;
+               if (i < db->salt_count) {
+                       s->next = ar[i].p;
+                       s = s->next;
+               }
        }
        s->next = 0;
```

**Comment by magnumripper** [comment]
Ouch! So it wasn't in the pot sync code at all. I'm puzzled that it only seems to affect pot sync and not normal cracking?

**Comment by claudioandre** [comment]
Me too. But, I guess the db->salt_hash[i] array is not really important in other situations? 

BTW: is this core (john proper)? Should I create a PR?

**Comment by magnumripper** [comment]
I committed it in 62fd99d. This is Jumbo specific, @jfoug wrote it. I'm not sure why it did not affect normal cracking, I will investigate further. Maybe it simply falls back to not using the salt hash list if the plaintext is not found.

Lots of thanks!

**Comment by magnumripper** [comment]
I wrote the "Rebuild salt hash table, if we still had one" part, which is pot-sync specific though... I guess @jfoug's original code worked fine but when I added that thing, I introduced the bug. So this should be pot sync specific. I think in John Proper the salt hash table is dropped after loading, but Jumbo now keeps it for quick pot sync. During normal cracking, JtR "knows" which salt it is currently processing, but within pot sync, any salt can come up.
